File: budget/frontend/src/lib/money.ts
/**
 * Rechnen mit ganzzahligen Minoreinheiten (Rappen/Cent).
 *
 * Diese Datei ist die Frontend-Entsprechung von `backend/app/services/money.py`.
 * Beide muessen bei identischer Eingabe identisch rechnen -- die Aufteilungs-Vorschau
 * im Erfassungsformular soll sofort das zeigen, was der Server anschliessend speichert.
 * Deshalb ist `allocate` hier bewusst dupliziert statt per Request geholt.
 */

export const MINOR_DIGITS = 2;
const MINOR_FACTOR = 10 ** MINOR_DIGITS;

/**
 * Verteilt `totalMinor` gewichtet und ohne Rundungsverlust (Verfahren des groessten
 * Rests). Bei gleichen Resten gewinnt die fruehere Position, damit der Rundungsrest
 * bei gleichmaessiger Verteilung an die erste Person geht.
 *
 * Es gilt immer: `allocate(t, w).reduce((a, b) => a + b, 0) === t`.
 */
export function allocate(totalMinor: number, weights: number[]): number[] {
  if (weights.length === 0) throw new Error("Mindestens ein Gewicht erforderlich");
  if (weights.some((w) => w < 0)) throw new Error("Gewichte duerfen nicht negativ sein");
  const totalWeight = weights.reduce((a, b) => a + b, 0);
  if (totalWeight <= 0) throw new Error("Die Summe der Gewichte muss groesser als 0 sein");

  const sign = totalMinor < 0 ? -1 : 1;
  const amount = Math.abs(totalMinor);

  const base = weights.map((w) => Math.floor((amount * w) / totalWeight));
  const remainders = weights.map((w, index) => ({
    rest: (amount * w) % totalWeight,
    index,
  }));

  let missing = amount - base.reduce((a, b) => a + b, 0);
  remainders.sort((a, b) => b.rest - a.rest || a.index - b.index);
  for (const { index } of remainders) {
    if (missing <= 0) break;
    base[index] += 1;
    missing -= 1;
  }

  return base.map((value) => sign * value);
}

/**
 * Liest eine Benutzereingabe als Minoreinheiten.
 * Akzeptiert `1234.50`, `1'234.50`, `1 234,50`, `12.-` und einfache Rechnungen
 * wie `12.50+3` (praktisch beim Erfassen einer Quittung).
 * Gibt `null` zurueck, wenn die Eingabe kein Betrag ist.
 */
export function parseAmountInput(input: string): number | null {
  const raw = input.trim();
  if (!raw) return null;

  const cleaned = raw
    // Geschütztes und schmales geschütztes Leerzeichen kommen als Tausendertrenner vor.
    .replace(/['’\s\u00A0\u202F]/g, "")
    .replace(/(CHF|EUR|€|\$)/gi, "")
    .replace(/\.-$/, "");

  if (/^[-+]?[\d.,]+([+\-*][\d.,]+)+$/.test(cleaned)) {
    // Kleine Kettenrechnung: Summanden einzeln lesen und in Minoreinheiten addieren.
    const parts = cleaned.match(/[+\-*]?[\d.,]+/g);
    if (!parts) return null;
    let total = 0;
    for (const part of parts) {
      const operator = /^[+\-*]/.test(part) ? part[0] : "+";
      const value = toMinor(part.replace(/^[+\-*]/, ""));
      if (value === null) return null;
      if (operator === "+") total += value;
      else if (operator === "-") total -= value;
      else total = Math.round((total * value) / MINOR_FACTOR);
    }
    return total;
  }

  return toMinor(cleaned);
}

function toMinor(text: string): number | null {
  let normalized = text;
  const hasComma = normalized.includes(",");
  const hasDot = normalized.includes(".");
  if (hasComma && hasDot) {
    normalized =
      normalized.lastIndexOf(",") > normalized.lastIndexOf(".")
        ? normalized.replace(/\./g, "").replace(",", ".")
        : normalized.replace(/,/g, "");
  } else if (hasComma) {
    normalized = normalized.replace(",", ".");
  }
  if (!/^[-+]?\d*(\.\d*)?$/.test(normalized) || !/\d/.test(normalized)) return null;

  const negative = normalized.startsWith("-");
  const [whole, fraction = ""] = normalized.replace(/^[-+]/, "").split(".");
  const padded = (fraction + "00").slice(0, MINOR_DIGITS + 1);
  const rounded = Math.round(Number(`${whole || "0"}.${padded}`) * MINOR_FACTOR);
  return negative ? -rounded : rounded;
}

/** Minoreinheiten als einfacher Dezimalstring, z. B. fuer Eingabefelder. */
export function toDecimalString(amountMinor: number): string {
  const sign = amountMinor < 0 ? "-" : "";
  const abs = Math.abs(amountMinor);
  return `${sign}${Math.floor(abs / MINOR_FACTOR)}.${String(abs % MINOR_FACTOR).padStart(MINOR_DIGITS, "0")}`;
}

/** Anteil als Prozentwert, oder `null` wenn der Nenner 0 ist. */
export function share(part: number, whole: number): number | null {
  if (whole === 0) return null;
  return (part / whole) * 100;
}

File: budget/backend/app/services/money.py
"""Rechnen mit ganzzahligen Minoreinheiten (Rappen/Cent).

Es gibt in dieser Anwendung keine Float-Betraege. Jede Verteilung eines Betrags auf
mehrere Personen laeuft ueber :func:`allocate`, damit die Summe der Teile immer exakt
dem Ausgangsbetrag entspricht.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal, InvalidOperation

MINOR_DIGITS = 2
_MINOR_FACTOR = 10**MINOR_DIGITS


def allocate(total_minor: int, weights: list[int]) -> list[int]:
    """Verteilt ``total_minor`` gewichtet und ohne Rundungsverlust.

    Verfahren: groesster Rest (largest remainder). Bei gleichen Resten gewinnt die
    fruehere Position -- dadurch landet der Rundungsrest bei gleichmaessiger Verteilung
    automatisch bei der ersten Person, wie in der Spezifikation gefordert.

    ``sum(allocate(t, w)) == t`` gilt fuer jedes ``t`` und jede Gewichtung.
    """
    if not weights:
        raise ValueError("Mindestens ein Gewicht erforderlich")
    if any(w < 0 for w in weights):
        raise ValueError("Gewichte duerfen nicht negativ sein")
    total_weight = sum(weights)
    if total_weight <= 0:
        raise ValueError("Die Summe der Gewichte muss groesser als 0 sein")

    sign = -1 if total_minor < 0 else 1
    amount = abs(total_minor)

    base: list[int] = []
    remainders: list[tuple[int, int]] = []  # (rest, index)
    for index, weight in enumerate(weights):
        share = amount * weight
        base.append(share // total_weight)
        remainders.append((share % total_weight, index))

    missing = amount - sum(base)
    # Groesster Rest zuerst, bei Gleichstand die kleinere Position.
    remainders.sort(key=lambda item: (-item[0], item[1]))
    for rest, index in remainders[:missing]:
        del rest
        base[index] += 1

    return [sign * value for value in base]


def parse_amount(value: str | int | float | Decimal) -> int:
    """Wandelt eine Benutzereingabe in Minoreinheiten.

    Akzeptiert ``1234.50``, ``1'234.50``, ``1 234,50`` und ``-12.-``.
    Floats werden ueber ihre String-Darstellung geleitet, damit keine Binaerartefakte
    einfliessen.
    """
    if isinstance(value, int):
        return value * _MINOR_FACTOR
    if isinstance(value, Decimal):
        decimal_value = value
    else:
        text = str(value).strip()
        if not text:
            raise ValueError("Leerer Betrag")
        text = text.replace("'", "").replace("’", "").replace(" ", "").replace(" ", "")
        text = text.replace("CHF", "").replace("EUR", "").replace("€", "")
        if "," in text and "." in text:
            # Letztes Trennzeichen ist das Dezimaltrennzeichen.
            if text.rfind(",") > text.rfind("."):
                text = text.replace(".", "").replace(",", ".")
            else:
                text = text.replace(",", "")
        elif "," in text:
            text = text.replace(",", ".")
        if text.endswith(".-"):
            text = text[:-2]
        try:
            decimal_value = Decimal(text)
        except InvalidOperation as exc:
            raise ValueError(f"Kein gueltiger Betrag: {value!r}") from exc
    quantized = decimal_value.quantize(Decimal(1).scaleb(-MINOR_DIGITS), rounding=ROUND_HALF_UP)
    return int(quantized.scaleb(MINOR_DIGITS))


def to_decimal(amount_minor: int) -> Decimal:
    return (Decimal(amount_minor) / _MINOR_FACTOR).quantize(Decimal(1).scaleb(-MINOR_DIGITS))


def format_amount(amount_minor: int) -> str:
    """Reine Punkt-Darstellung fuer CSV-Export. Anzeige-Formatierung macht das Frontend."""
    return f"{to_decimal(amount_minor):.{MINOR_DIGITS}f}"
